File: src/main/java/com/kasisoft/libs/common/functional/KBiPredicate.java
package com.kasisoft.libs.common.functional;

import com.kasisoft.libs.common.*;

import jakarta.validation.constraints.*;

import java.util.function.*;

/**
 * @author daniel.kasmeroglu@kasisoft.com
 */
@FunctionalInterface
public interface KBiPredicate<T, U> {

    boolean test(T arg1, U arg2) throws Exception;

    @NotNull
    default KBiPredicate<T, U> and(@NotNull KBiPredicate<? super T, ? super U> other) {
        return ($t, $u) -> test($t, $u) && other.test($t, $u);
    }

    @NotNull
    default KBiPredicate<T, U> negate() {
        return ($t, $u) -> !test($t, $u);
    }

    @NotNull
    default KBiPredicate<T, U> or(@NotNull KBiPredicate<? super T, ? super U> other) {
        return ($t, $u) -> test($t, $u) || other.test($t, $u);
    }

    @NotNull
    default BiPredicate<T, U> protect() {
        return ($t, $u) -> {
            try {
                return test($t, $u);
            } catch (Exception ex) {
                throw KclException.wrap(ex);
            }
        };
    }

} /* ENDINTERFACE */

File: src/main/kotlin/com/kasisoft/libs/common/functional/KBiPredicate.kt
package com.kasisoft.libs.common.functional;

import com.kasisoft.libs.common.*;

import java.util.function.*;

/**
 * This implementation allows to throw Exceptions (useful on the Java side)
 *
 * @author daniel.kasmeroglu@kasisoft.com
 */
fun interface KBiPredicate<T: Any?, U: Any?> {

    @Throws(Exception::class)
    fun test(t: T, u: U): Boolean

    fun protect(): BiPredicate<T, U> =
        BiPredicate<T, U> { t: T, u: U ->
            try {
                test(t, u)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

    fun negate(): BiPredicate<T, U> =
        BiPredicate { t: T, u: U ->
            try {
                !test(t, u)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

    fun and(after: BiPredicate<T, U>) =
        BiPredicate { t: T, u: U ->
            try {
                test(t, u) && after.test(t, u)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

    fun and(after: KBiPredicate<T, U>) =
        BiPredicate { t: T, u: U ->
            try {
                test(t, u) && after.test(t, u)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

    fun or(after: BiPredicate<T, U>) =
        BiPredicate { t: T, u: U ->
            try {
                test(t, u) || after.test(t, u)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

    fun or(after: KBiPredicate<T, U>) =
        BiPredicate { t: T, u: U ->
            try {
                test(t, u) || after.test(t, u)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

} /* ENDINTERFACE */
